### PR TITLE
Add validation and support for non-default address space in global variables, and fix tests by adding explicit address space

### PR DESCRIPTION
**Pull request by VyacheslavLevytskyy**
Global variables in LLVM IR with the default address space (Private) are translated into invalid SPIR-V where OpVariable is created with a storage class of Function. The problem postponed to the next step of validation where SPIR-V validator tool detects problematic SPIR-V code from translator and reports it as the following: "Variables can not have a function[7] storage class outside of a function".

This PR is to address the issue, including:
- ensuring that global variable can not have Function storage class
- ensuring that intrinsic global variables `llvm.global_ctors` and `llvm.global_dtors` in LLVM IR when defined with non-default address space will be converted to SPIRV and back to LLVM IR in a correct way, so that "appending" linkage is added for them
- fixing tests related to the issue

### CODE REVIEW

**Comment by VyacheslavLevytskyy** [comment]
FYI @MrSidims

File: test/extensions/EXT/SPV_EXT_relaxed_printf_string_address_space/non-constant-printf.ll [hunk 1]
```diff
@@ -38,14 +49,16 @@
 target datalayout = "e-p:32:32:32-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v24:32:32-v32:32:32-v48:64:64-v64:64:64-v96:128:128-v128:128:128-v192:256:256-v256:256:256-v512:512:512-v1024:1024:1024"
 target triple = "spir-unknown-unknown"
 
-@0 = internal unnamed_addr addrspace(0) constant [6 x i8] c"Test\0A\00", align 1
+@0 = internal unnamed_addr addrspace(2) constant [6 x i8] c"Test\0A\00", align 1
```
Comment (asudarsa) [inline test/extensions/EXT/SPV_EXT_relaxed_printf_string_address_space/non-constant-printf.ll:52]: Why is this change required?

Thanks

File: test/extensions/INTEL/SPV_INTEL_fpga_memory_attributes/IntelFPGAMemoryAttributesForStaticVar.ll [hunk 2]
```diff
@@ -66,15 +66,16 @@ target triple = "spir"
...
+@.str.4 = private unnamed_addr addrspace(1) constant [49 x i8] c"{memory:DEFAULT}{sizeinfo:4}{force_pow2_depth:0}\00", section "llvm.metadata"
+;;;@llvm.global.annotations = appending addrspace(1) global [4 x { ptr, ptr, ptr, i32 }] [{ ptr, ptr, ptr, i32 } { ptr addrspacecast (ptr addrspace(1) @_ZZ13numbanks_statiE5a_one to ptr), ptr addrspace(1) @.str, ptr addrspace(1) @.str.1, i32 2 }, { ptr, ptr, ptr, i32 } { ptr addrspacecast (ptr addrspace(1) @_ZZ11memory_statcE5b_one to ptr), ptr addrspace(1) @.str.2, ptr addrspace(1) @.str.1, i32 7 }, { ptr, ptr, ptr, i32 } { ptr addrspacecast (ptr addrspace(1) @_ZZ13annotate_statiE5c_one to ptr), ptr addrspace(1) @.str.3, ptr addrspace(1) @.str.1, i32 12 }, { ptr, ptr, ptr, i32 } { ptr addrspacecast (ptr addrspace(1) @_ZZ21force_pow2_depth_statiE9fp2d_stat to ptr), ptr addrspace(1) @.str.4, ptr addrspace(1) @.str.1, i32 17 }], section "llvm.metadata"
```
Comment (asudarsa) [inline test/extensions/INTEL/SPV_INTEL_fpga_memory_attributes/IntelFPGAMemoryAttributesForStaticVar.ll:77]: Is this line needed/

Thanks

File: test/llvm-intrinsics/instrprof.ll [hunk 3]
```diff
@@ -28,9 +28,11 @@ $handler = comdat any
 ; Function Attrs: convergent mustprogress norecurse
 define weak_odr dso_local spir_kernel void @handler() #0 comdat !kernel_arg_buffer_location !1 {
 entry:
-  call void @llvm.instrprof.increment(ptr @__profn__, i64 0, i32 1, i32 0)
-  call void @llvm.instrprof.increment.step(ptr @__profn__, i64 0, i32 1, i32 0, i64 0)
-  call void @llvm.instrprof.value.profile(ptr @__profn__, i64 0, i64 0, i32 1, i32 0)
+  %tmp1 = alloca [7 x i8]
+  call void @llvm.memcpy.p0.p1.i64(ptr %tmp1, ptr addrspace(1) @__profn__, i64 7, i1 false)
```
Comment (asudarsa) [inline test/llvm-intrinsics/instrprof.ll:32]: Why do we need this memcpy?
Sorry for ignorant question.

Thanks

**Review by asudarsa — COMMENTED** [review]
Very good set of changes. Thanks. A few questions. Please address.

**Comment by asudarsa** [comment]
Suggested change to the PR topic:
Add Validation and support for non-default address space in global variables and fix tests by adding explicit address space.

File: test/extensions/INTEL/SPV_INTEL_fpga_memory_attributes/IntelFPGAMemoryAttributesForStaticVar.ll (hunk 2 above)
Comment (VyacheslavLevytskyy) [reply test/extensions/INTEL/SPV_INTEL_fpga_memory_attributes/IntelFPGAMemoryAttributesForStaticVar.ll:77]: Thank you, it should be removed of course.

File: test/extensions/EXT/SPV_EXT_relaxed_printf_string_address_space/non-constant-printf.ll (hunk 1 above)
Comment (VyacheslavLevytskyy) [reply test/extensions/EXT/SPV_EXT_relaxed_printf_string_address_space/non-constant-printf.ll:52]: It's exactly the reason for creation of the whole PR. addrspace(0) is a default address space (Private) that is mapped to  Function storage class in SPIR-V, violating validation rules. In this case we need to create a global variable in other than 0 address space.

File: test/extensions/EXT/SPV_EXT_relaxed_printf_string_address_space/non-constant-printf.ll [hunk 4]
```diff
@@ -14,19 +14,30 @@
 
 ; CHECK-SPIRV: Extension "SPV_EXT_relaxed_printf_string_address_space"
 ; CHECK-SPIRV: ExtInstImport [[#ExtInstSetId:]] "OpenCL.std"
-; CHECK-SPIRV-DAG: TypeInt [[#TypeInt8Id:]] 8 0
 ; CHECK-SPIRV-DAG: TypeInt [[#TypeInt32Id:]] 32 0
-; CHECK-SPIRV-DAG: TypePointer [[#FunctionStorClassPtrTy:]] 7 [[#TypeInt8Id]]
-; CHECK-SPIRV-DAG: TypePointer [[#WGStorClassPtrTy:]] 5 [[#TypeInt8Id]]
-; CHECK-SPIRV-DAG: TypePointer [[#CrossWFStorClassPtrTy:]] 4 [[#TypeInt8Id]]
-; CHECK-SPIRV-DAG: TypePointer [[#GenericStorCalssPtrTy:]] 8 [[#TypeInt8Id]]
+; CHECK-SPIRV-DAG: TypeInt [[#TypeInt8Id:]] 8 0
```
[the remark this answers is not in the document]
Comment (VyacheslavLevytskyy) [reply test/extensions/EXT/SPV_EXT_relaxed_printf_string_address_space/non-constant-printf.ll:18]: A good point indeed, thank you. It's the only case when I took a liberty to modify a test checks slightly more than it was absolutely needed, and it was thought out before the introduced change. My motivation is that I've spent quite a long time analyzing SPIR-V code against previously existed checks until I map a logics behind them into SPIR-V lines, and understood that those checks took an arbitrary piece of SPIR-V code in the middle of the generated module, passing without a validation a previous very similar block. That is too subtle and not obvious for a reader of the test against the generated SPIR-V module. Thus, the intent of this additional checks is to follow by checks the expected structure and sequence of the generated SPIR-V module in a more obvious way, and to help the next person who will be working with this test in future to quicker map a structure of checks with the structure of expected SPIR-V module to generate. As an additional benefit from making the test more readable, we gain a more detailed validation of the generated module.

**Review by asudarsa — APPROVED** [review]
Thanks so much for detailed answers. I do agree that the tests are in good shape as is.

LGTM.

**Comment by bwlodarcz** [comment]
LGTM

File: test/GlobalVarNoAddrspace.ll [hunk 5]
```diff
@@ -0,0 +1,16 @@
+; RUN: llvm-as %s -o %t.bc
+; RUN: not llvm-spirv %t.bc 2>&1 \
+; RUN: | FileCheck %s --check-prefix=CHECK-ERROR
+
+; CHECK-ERROR: InvalidInstruction: Can't translate llvm instruction:
+; CHECK-ERROR-NEXT: Global variable can not have Function storage class. Consider setting a proper address space.
```
Comment (svenvh) [inline test/GlobalVarNoAddrspace.ll:6]: Would it make sense to move this test into `test/negative/`?

Comment (VyacheslavLevytskyy) [reply test/GlobalVarNoAddrspace.ll:6]: Thank you @svenvh We have started to disperse negative tests over the whole test suite, and tend to group it by subject rather than by the negative criteria.

Comment (svenvh) [reply test/GlobalVarNoAddrspace.ll:6]: This test doesn't appear to be grouped by subject as it's directly under `test/` (ideally we should aim to reduce the number of tests in the top-level directory and group them by subject as you say).

Comment (VyacheslavLevytskyy) [reply test/GlobalVarNoAddrspace.ll:6]: You are absolutely right, I've been thinking about this in the same terms when creating the PR. If it's a blocker, I will create another commit, grouping two tests GlobalVarNoAddrspace and GlobalVarAddrspace together in a new folder. However, it looked for me like a bit exaggeration to create a new folder for just two new tests.

You are right though that the top-level directory should be reordered by topics. May we plan this as a wider scope activity, and address the top-level directory next time, or you would prefer me to rework this PR and place those two tests into a new dedicated folder?

Comment (svenvh) [reply test/GlobalVarNoAddrspace.ll:6]: I agree that a directory with just 2 tests is a bit of overkill.

**Comment by VyacheslavLevytskyy** [comment]
@asudarsa Fail in a Windows build is unrelated to this PR, are we able to merge as it is now? Thank you

**Comment by asudarsa** [comment]
> @asudarsa Fail in a Windows build is unrelated to this PR, are we able to merge as it is now? Thank you

I agree it's unmerged. But it's also preventing the subsequent tests from running. So, we will need to get a successful test report before merging.

Thanks

**Comment by svenvh** [comment]
Closed/reopened to restart CI.

Before merging, I would like @MrSidims or @vmaksimo to have a look too.

**Comment by VyacheslavLevytskyy** [comment]
> Closed/reopened to restart CI.
> 
> Before merging, I would like @MrSidims or @vmaksimo to have a look too.

@svenvh I'm afraid @MrSidims is not available in December.

@vmaksimo May I ask you to have a look at the PR please?

File: test/extensions/INTEL/SPV_INTEL_function_pointers/global_ctor_dtor.ll [hunk 6]
```diff
@@ -73,4 +74,4 @@ attributes #0 = { nounwind sspstrong "correctly-rounded-divide-sqrt-fp-math"="fa
 !opencl.compiler.options = !{!1}
 
 !0 = !{i32 1, i32 2}
-!1 = !{}
+!1 = !{}
```
Comment (vmaksimo) [inline test/extensions/INTEL/SPV_INTEL_function_pointers/global_ctor_dtor.ll:77]: Nit: blank line should not be removed from the end of the file

**Review by vmaksimo — APPROVED** [review]
